### PR TITLE
redirect blob download to s3 server

### LINKED ISSUE #494: s3: add a 'redirect' flag in storageDriver params and based on it serve blobs or redirect to S3 URLs
**Is your feature request related to a problem? Please describe.**
When using s3 as backend storage zot first downloads blobs from s3 and then serves them thus wasting time.
It should be able to use redirects to s3 urls based on a 'redirect' flag in storageDriver config parameters.

**Describe the solution you'd like**

If redirect flag is true then:
1) use path := storage.BlobPath() to get path to blob
2) use url := storageDriver.URLFor(path)
3) in routes use http.Redirect(url) and serve the URLs instead of blobs.

we can add a new func to Storage interface URLForPath(path) that calls storageDriver.URLFor(path)
for filesystem storage return a err.NotImplemented.

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
**What type of PR is this?**

**Which issue does this PR fix**:
[Issue ](https://github.com/project-zot/zot/issues/494)#494
Closes #494 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/routes.go [hunk 1]
```diff
@@ -621,6 +622,26 @@ func (rh *RouteHandler) GetBlob(response http.ResponseWriter, request *http.Requ
...
+			http.StatusBadRequest,
+			NewErrorList(NewError(DIGEST_INVALID, map[string]string{"digest": digest})))
+	}
+
+	if rh.c.Config.Storage.StorageDriver != nil {
+		redirect, ok := rh.c.Config.Storage.StorageDriver["redirect"].(bool)
+
+		if ok && redirect {
+			path := imgStore.BlobPath(name, digestObj)
+			if path, err := imgStore.URLForPath(path); err == nil {
```
Comment (peusebiu) [inline pkg/api/routes.go:637]: please log the error in case there is one :D, it could be. and return a http.StatusInternalServerError

Comment (laurentiuNiculae) [reply pkg/api/routes.go]: I've addressed this comment. I'll pay more attention to these details :D

**Review by peusebiu — DISMISSED** [review]
LGTM

**Comment by andaaron** [comment]
- Need to check if clients handle http redirects on a common basis. Need to check if OCI dist spec mentions redirects.
- After we validate the above, we need a rebase

**Comment by rchincha** [comment]
@laurentiuNiculae is this PR still valid, else we should close it.

**Comment by KuxaBeast** [comment]
Is this still relevant? I am looking forward to this feature... Is there any reason why haven't this PR been merged yet?

**Comment by rchincha** [comment]
> Is this still relevant? I am looking forward to this feature... Is there any reason why haven't this PR been merged yet?

@KuxaBeast, we plan to merge this PR in a future milestone. Meanwhile, could you give us more details about your use case?

**Comment by KuxaBeast** [comment]
@rchincha, it's just a home-lab setup for now. I was thinking about making use of the Backblaze storage and Cloudflare, utilizing the Bandwidth Alliance, making the download costs essentially zero. But then I found that it might also be a good idea to offload downloads directly to Cloudflare, instead of bombing the Zot instance on my K8s cluster.

**Pull request closed without merging**